### ISSUE TITLE
Misidentified disk size

### ISSUE THREAD

**Issue by FlyingRatBull** [veracrypt/VeraCrypt#1180]
# Expected behavior
Creating a hidden partition of about *4TiB* size is possible on a *4TiB* drive.

# Observed behavior
Only a partition of *416MiB* is possible.

# Steps to reproduce
1. Checking the disks size:
```console
foo@bar:~$ sudo fdisk -l /dev/disk/by-id/MY-DISK
Disk /dev/disk/by-id/MY-DISK: 3.64 TiB, 4000787030016 bytes, 7814037168 sectors
```
2. Trying to create the hidden partition:
```console
foo@bar:~$ sudo veracrypt -t -c --volume-type=hidden /dev/disk/by-id/MY-DISK --encryption=twofish --hash=whirlpool --filesystem=none -p 'mypassword' --pim=0 -k "" --random-source=/dev/urandom --size=3950G
Error: Incorrect volume size
```
3. Using the interactive variant:
```console
foo@bar:~$ sudo veracrypt -t -c --volume-type=hidden /dev/disk/by-id/MY-DISK --encryption=twofish --hash=whirlpool --filesystem=none -p 'mypassword' --pim=0 -k "" --random-source=/dev/urandom
Enter hidden volume size (sizeK/size[M]/sizeG/sizeT/max): 3950G
Error: Maximum volume size is 416 MiB.
```

# Screenshots

# Your Environment

Please tell us more about your environment

**VeraCrypt version:** 1.25.9

**Operating system and version:** Ubuntu 22.04.2 LTS

**System type:** 64bit

**Comment by Jertzukka** [comment]
If you have a Windows device, you could check in "Direct Mode" whether it gives the same largest possible size for the hidden volume as the Linux CLI to rule out platform/cli issues.

What is the filesystem on your outer volume? For VeraCrypt to get the most accurate reading of the available space for hidden volume, I'd recommend exFAT for the outer volume. NTFS, ext4 and some of the others have different properties (due to journaling I believe) due to which the space left for hidden volume is way less.

**Comment by FlyingRatBull** [comment]
> If you have a Windows device, you could check in "Direct Mode" whether it gives the same largest possible size for the hidden volume as the Linux CLI to rule out platform/cli issues.

I unfortunately don't have a Windows installation.

> What is the filesystem on your outer volume? For VeraCrypt to get the most accurate reading of the available space for hidden volume, I'd recommend exFAT for the outer volume. NTFS, ext4 and some of the others have different properties (due to journaling I believe) due to which the space left for hidden volume is way less.

The outer filesystem is ZFS. Could this be the problem?

**Comment by Jertzukka** [comment]
Unfortunately I don't have experience how ZFS behaves, also you probably had to create the outer volume without a filesystem and manually format it with ZFS as I don't see it being on the filesystem options list in `VolumeCreator.h`. It's possible that it works fine and nobody has just taken the time to add it, or that someone has tested it and it didn't behave well (though I didn't see any other issues related to ZFS.)

**Comment by idrassi** [comment]
Actually there was a logical bug in the method `TextUserInterface::CreateVolume `that happend only in the case of CLI creation of a hidden volume on a device: the code was trying to get maximum available space from the "`/dev`" filesystem instead of using the real device size.

I have fixed this in the commit https://github.com/veracrypt/VeraCrypt/commit/9d8afdad9342c3e06e796c6b02eadb7f12cb6611.

@FlyingRatBull I will publish Linux binaries for 1.26.5 that contain this fix in few hours. Until then, you can also build VeraCrypt from source code using master branch.

**Comment by idrassi** [comment]
@FlyingRatBull I have uploaded .deb package for Ubuntu 22.04 to Sourceforge [Nightly Builds folders](https://sourceforge.net/projects/veracrypt/files/VeraCrypt%20Nightly%20Builds/Linux/). You can grab it from there.

This issue has been around forever and unless I missed an already opened issue for it, this seems to be the first time it is reported.

**Comment by FlyingRatBull** [comment]
This fixed it, it now detected a size of 3.6 TiB. Thank you for your quick help!
